File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/repository/heap/JavaClassRepository.java
package com.github.ol_loginov.heaplibweb.repository.heap;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;
import java.util.stream.Stream;

@Repository
public interface JavaClassRepository extends CrudRepository<JavaClassEntity, JavaClassEntity.PK> {
	Optional<JavaClassEntity> findByHeapIdAndName(int heapId, String name);

	@Query(value = "select * from JavaClass where heapId = :heapId and regexp_like(name, :nameRegex)", nativeQuery = true)
	Stream<JavaClassEntity> findAllByHeapIdAndNameMatches(int heapId, String nameRegex);

	Stream<JavaClassEntity> streamAllByHeapId(int heapId);

	Stream<JavaClassEntity> streamAllByHeapIdAndSuperClassId(int heapId, long superClassId);
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/repository/heap/JavaClassRepository.kt
package com.github.ol_loginov.heaplibweb.repository.heap

import java.util.stream.Stream

interface JavaClassRepository {
    fun persist(entity: JavaClassEntity)
    fun persistAll(entities: List<JavaClassEntity>)

    fun updateCounts(list: List<Pair<ULong, Int>>)

    fun findById(id: Long): JavaClassEntity?
    fun findByIdSure(id: Long): JavaClassEntity
    fun findByName(name: String): JavaClassEntity?

    /**
     * select * from JavaClass where heapId = :heapId and regexp_like(name, :nameRegex)
     */
    fun findAllByNameRegex(nameRegex: String): Stream<JavaClassEntity>
    fun streamAll(): Stream<JavaClassEntity>
    fun streamAllBySuperClassId(superClassId: Long): Stream<JavaClassEntity>
}